### PR TITLE
USM: refactor GoTLS

### PR DESCRIPTION
**Pull request by Yumasi**
### What does this PR do?

Port our GoTLS implementation to the new USM monitor architecture.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/usm/ebpf_gotls.go [hunk 1]
```diff
@@ -156,22 +157,53 @@ type GoTLSProgram struct {
...
+	Factory: newGoTLSProgram,
+	Maps: []*manager.Map{
+		{Name: offsetsDataMap},
+		{Name: goTLSReadArgsMap},
+		{Name: goTLSWriteArgsMap},
+	},
+	Probes: []*manager.Probe{
+		{
+			ProbeIdentificationPair: manager.ProbeIdentificationPair{
+				EBPFFuncName: "uprobe__crypto_tls_Conn_Read",
```
Comment (guyarb) [inline pkg/network/usm/ebpf_gotls.go:170]: use the consts

File: pkg/network/usm/ebpf_gotls.go [hunk 2]
```diff
@@ -191,60 +223,28 @@ func newGoTLSProgram(c *config.Config) *GoTLSProgram {
...
-
-func (p *GoTLSProgram) Start() {
+func (p *goTLSProgram) PreStart(m *manager.Manager) error {
 	var err error
-	p.offsetsDataMap, _, err = p.manager.GetMap(offsetsDataMap)
+	p.offsetsDataMap, _, err = m.GetMap(offsetsDataMap)
 	if err != nil {
 		log.Errorf("could not get offsets_data map: %s", err)
-		return
+		return err
```
Comment (guyarb) [inline pkg/network/usm/ebpf_gotls.go:247]: return the error instead of logging and returning an error

File: pkg/network/usm/ebpf_main.go [hunk 3]
```diff
@@ -133,45 +96,25 @@ func newEBPFProgram(c *config.Config, connectionProtocolMap *ebpf.Map, bpfTeleme
...
-		tailCallRouter:        tailCalls,
 		connectionProtocolMap: connectionProtocolMap,
 	}
 
 	return program, nil
 }
 
 func (e *ebpfProgram) Init() error {
-	var undefinedProbes []manager.ProbeIdentificationPair
+	undefinedProbes := make([]manager.ProbeIdentificationPair, 0)
```
Comment (guyarb) [inline pkg/network/usm/ebpf_main.go:109]:
```suggestion
	undefinedProbes := make([]manager.ProbeIdentificationPair, 0, len(e.tailCallRouter))
```

File: pkg/network/usm/ebpf_gotls.go [hunk 4]
```diff
@@ -191,60 +223,28 @@ func newGoTLSProgram(c *config.Config) *GoTLSProgram {
...
 }
 
-func (p *GoTLSProgram) Name() string {
+func (p *goTLSProgram) Name() string {
 	return "go-tls"
 }
 
-func (p *GoTLSProgram) IsBuildModeSupported(mode buildMode) bool {
-	return mode == CORE || mode == RuntimeCompiled
-}
```
[the remark this answers is not in the document]
Comment (Yumasi) [reply pkg/network/usm/ebpf_gotls.go:203]: That is actually more tricky that it first seems. I pushed a fix for this, which involves adding the build mode as an argument to `PreStart` (that is as close as it was before in terms of behaviour). I am not 100% if it is the best solution, so I'd love to have your opinion on this.

**Review by guyarb — APPROVED** [review]

**Comment by guyarb** [comment]
The last commit shouldn't affect the failed tests, can you check it again?

**Pull request merged**